cmd/kubeadm/app/util/apiclient/idempotency.go

```diff
@@ -51,6 +54,46 @@ func CreateOrUpdateConfigMap(client clientset.Interface, cm *v1.ConfigMap) error
 	return nil
 }
 
+// CreateOrMutateConfigMap tries to create the ConfigMap provided as cm. If the resource exists already, the latest version will be fetched from
+// the cluster and cmmutator callback will be called on it, then an Update of the mutated ConfigMap will be performed. This function is resilient
+// to conflicts, and a retry will be issued if the configmap was modified on the server between the refresh and the update (while the mutation was
+// taking place)
+func CreateOrMutateConfigMap(client clientset.Interface, cm *v1.ConfigMap, cmmutator ConfigMapMutator) error {
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(cm); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "unable to create configmap")
+		}
+		return MutateConfigMap(client, metav1.ObjectMeta{Namespace: cm.ObjectMeta.Namespace, Name: cm.ObjectMeta.Name}, cmmutator)
+	}
+	return nil
+}
+
+// MutateConfigMap takes a ConfigMap Object Meta (namespace and name), retrieves the resource from the server and tries to mutate it
+// by calling to the cmmutator callback, then an Update of the mutated ConfigMap will be performed. This function is resilient
+// to conflicts, and a retry will be issued if the configmap was modified on the server between the refresh and the update (while the mutation was
+// taking place)
+func MutateConfigMap(client clientset.Interface, cmmeta metav1.ObjectMeta, cmmutator ConfigMapMutator) error {
+	for {
```

Comment:
What are the max retries and timeout on this?

Reply:
Updated to use `k8s.io/client-go/util/retry`

Reply:
we might want to use `Steps: 0` but with a sensibly large cap e.g. `Cap: 1 * time.Minute`:
https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Backoff

Reply:
I'd argue that someone adding 100 control plane nodes is doing something massively wrong.

Reply:
@neolit123 `Steps: 0` won't work because of https://github.com/kubernetes/apimachinery/blob/b874eabb9a4eb99cef27db5c8d06f16542580cec/pkg/util/wait/wait.go#L283, the condition will never be executed.

Reply:
i guess that's true. in that case we need to see what number of steps and cap we need to settle on. what is the retry rate you are seeing, btw?

Reply:
@neolit123 all my tests with this are very artificial. I can try to do some real testing on this with e.g. 1 master + 4 concurrent joins, however it's also tricky to force the configmap update race on a complete join process, and running `kubeadm config upload from-file` again seems artificial to me.

I think it makes sense to start with some defaults that we consider sane and learn with the experience, sorry but I don't have a better answer since I believe giving a good enough answer requires some explicit efforts here.

Reply:
i think the artificial tests are OK as a base for the estimate here.
after all, we are only interested in the config map update.

so if a unit test passes for a excess number of concurrent clients, the real world case should be good too. at least that;'s how i see it.

we might be able to test this in e2e eventually too.

Reply:
> so if a unit test passes for a excess number of concurrent clients, the real world case should be good too. at least that;'s how i see it.

Oh, you propose to start an apiserver for the unit tests? I have been using the fake client (that cannot mimic real server-side conflicts, it's just reactors that can mimic a conflict returned by the server, but it's not testing any real concurrency). I can look into that, but as far as I remember starting an apiserver was flaky and in general not desired, I might be wrong though.

Reply:
yeah, a real server is bad idea. it flakes in parallel unit tests; there are also endpoint and port conflicts etc. i was hopping that there is a fake api server that can be used here..

alternatively, we can have this as a test in test/e2e-kubeadm where a set of concurrent clients are trying to update the real ClusterStatus ConfigMap of the real server that the suite is being run on.

Reply:
I think e2e tests are better for this purpose, because with a test server we would have to mimic ourselves a basic conflict return (version handling and so on); all in memory, but still, I think it will include some boilerplate code that we'll have to maintain in the end.

I will look into the e2e tests and see if they can be improved to include a multi-writer test. Do you want me to tackle that as part of this PR as well?

Reply:
thanks. i think a separate PR is better.

@fabriziopandini merged a change recently that allows you to run this kubeadm suite on an existing cluster like so: `make test-e2e-kubeadm`
https://github.com/kubernetes/kubernetes/pull/75546

Reply:
I will work on that as a separate PR as soon as this one merges. Thank you!